frame/support/src/dispatch.rs

```diff
@@ -3496,111 +3496,177 @@ mod tests {
-	}
+		#[pallet::config]
+		#[pallet::disable_frame_system_supertrait_check]
+		pub trait Config: 'static {
+			type BlockNumber: Parameter + Default + MaxEncodedLen;
+			type AccountId;
+			type Balance;
+			type BaseCallFilter: crate::traits::Contains<Self::RuntimeCall>;
+			type RuntimeOrigin;
+			type RuntimeCall;
+			type PalletInfo: crate::traits::PalletInfo;
+			type DbWeight: Get<crate::weights::RuntimeDbWeight>;
+		}
 
-	decl_module! {
-		pub struct Module<T: Config> for enum Call where origin: T::RuntimeOrigin, system=self {
+		#[pallet::error]
+		pub enum Error<T> {
+			/// Required by construct_runtime
+			CallFiltered,
+		}
+
+		#[pallet::origin]
+		pub type Origin<T> = RawOrigin<<T as Config>::AccountId>;
+
+		#[pallet::call]
+		impl<T: Config> Pallet<T> {
 			// no arguments, fixed weight
-			#[weight = 1000]
-			fn f00(_origin) { unimplemented!(); }
+			#[pallet::weight(1000)]
+			pub fn f00(_origin: OriginFor<T>) -> DispatchResult {
+				unimplemented!();
+			}
 
-			#[weight = (1000, DispatchClass::Mandatory)]
-			fn f01(_origin) { unimplemented!(); }
+			#[pallet::weight((1000, DispatchClass::Mandatory))]
+			pub fn f01(_origin: OriginFor<T>) -> DispatchResult {
+				unimplemented!();
+			}
 
-			#[weight = (1000, Pays::No)]
-			fn f02(_origin) { unimplemented!(); }
+			#[pallet::weight((1000, Pays::No))]
+			pub fn f02(_origin: OriginFor<T>) -> DispatchResult {
+				unimplemented!();
+			}
 
-			#[weight = (1000, DispatchClass::Operational, Pays::No)]
-			fn f03(_origin) { unimplemented!(); }
+			#[pallet::weight((1000, DispatchClass::Operational, Pays::No))]
+			pub fn f03(_origin: OriginFor<T>) -> DispatchResult {
+				unimplemented!();
+			}
 
 			// weight = a x 10 + b
-			#[weight = ((_a * 10 + _eb * 1) as u64, DispatchClass::Normal, Pays::Yes)]
-			fn f11(_origin, _a: u32, _eb: u32) { unimplemented!(); }
+			#[pallet::weight(((_a * 10 + _eb * 1) as u64, DispatchClass::Normal, Pays::Yes))]
+			pub fn f11(_origin: OriginFor<T>, _a: u32, _eb: u32) -> DispatchResult {
+				unimplemented!();
+			}
 
-			#[weight = (0, DispatchClass::Operational, Pays::Yes)]
-			fn f12(_origin, _a: u32, _eb: u32) { unimplemented!(); }
+			#[pallet::weight((0, DispatchClass::Operational, Pays::Yes))]
+			pub fn f12(_origin: OriginFor<T>, _a: u32, _eb: u32) -> DispatchResult {
+				unimplemented!();
+			}
 
-			#[weight = T::DbWeight::get().reads(3) + T::DbWeight::get().writes(2) + Weight::from_ref_time(10_000)]
-			fn f20(_origin) { unimplemented!(); }
+			#[pallet::weight(T::DbWeight::get().reads(3) + T::DbWeight::get().writes(2) + Weight::from_ref_time(10_000))]
+			pub fn f20(_origin: OriginFor<T>) -> DispatchResult {
+				unimplemented!();
+			}
+
+			#[pallet::weight(T::DbWeight::get().reads_writes(6, 5) + Weight::from_ref_time(40_000))]
+			pub fn f21(_origin: OriginFor<T>) -> DispatchResult {
+				unimplemented!();
+			}
+		}
+
+		pub mod pallet_prelude {
+			pub type OriginFor<T> = <T as super::Config>::RuntimeOrigin;
```

Comment:
Couldn't you have just imported this type alias from `frame_system::pallet_prelude`?

Reply:
No, `frame-support` is a dependency of `frame-system`, I don't wanna import the `frame-system` as a dev-dependency of `frame-support`.
It can easily cause confusion between existing `frame-system` modules and the module `frame-system` for tests.

Reply:
But `frame-system` is already a dev dependency for `frame-support`:

https://github.com/paritytech/substrate/blob/37664fe5b3513eb996225f016eceaf74963b8133/frame/support/Cargo.toml#L46

Reply:
It's only used for doc tests of `decl_module`. I will remove it from `dev-dependency` of `frame-support`.

Reply:
What sort of confusion are you really referring to? Since we've already been depending on `frame-system`, there shouldn't be any problems to continue depending upon it in `frame-support`?

Reply:
if `frame-system crate` is introduced as dev-dependency , it will easily conflict with our custom `frame-system module` in use.

Reply:
It seems that it‘s not possible to remove `frame-system` from dependencies or dev-dependencies of `frame-support` currently. The error message is as follows:

```
error: Could not find `frame-system` in `dependencies` or `dev-dependencies` in `/home/koushiro/Code/parity/substrate/frame/support/Cargo.toml`!
  --> frame/support/src/storage/generator/mod.rs:48:2
   |
48 |     #[crate::pallet]
   |     ^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the attribute macro `crate::pallet` (in Nightly builds, run with -Z macro-backtrace for more info)
```

For unit tests of `frame-support`, I prefer not to use `frame-system` directly in tests, maybe we can use `frame-system` in integration tests (`frame-support-test`)? What do you think?

Reply:
> I mean that if `frame-system crate` is introduced as dev-dependency , it will easily conflict with our custom `frame-system module` in use.

I don't understand this at all. We could just specify `crate::frame_system` and `frame_system`. The macros themselves should also have ensured that they refer to the correct `frame_system` during expansion, and if they don't, then it's a bug.

Reply:
I'd like to reiterate that I think `frame-support`'s unit tests should not introduce `frame-system` (because `frame-system` itself is dependent on `frame-support`), and the `pallet` macros impose restrictions on the name  and crate dependencies of the lowest level module (the lowest-level module name must be `frame-system`, and crate dependencies must contain `frame-system`)

Reply:
Look, we already heard what you said, and the reasoning you gave doesn't make any sense, _especially_ when it is already being depended upon without your changes. As I said before, if the macros are imposing changes, then let's take a closer look at what restrictions they're actually imposing and modify them accordingly.

Reply:
We cannot use `OriginFor` from `frame-system` here since `T` does not implement `frame_system::Config` (see the `disable_frame_system_supertrait_check`).

Furthermore i think this line is somehow buggy… https://github.com/paritytech/substrate/blob/5819373b70d2d9665821948f2913c6d8d9019090/frame/support/procedural/src/pallet/expand/call.rs#L364

It tries to find `pallet_prelude` in the local pallet named `frame_system` and not from the proper `frame_system` crate.

Reply:
That doesn't look like to be the case -- `#frame_system` looks like it is properly looking for the `frame_system` pallet, and not simply looking for a local crate named `frame_system`.

Furthermore, there is actually a `pallet_prelude` module in the `frame_system` crate:

https://github.com/paritytech/substrate/blob/c7d26dcf3d6c19fa7dcbd1f2633e337e5a5cbc40/frame/system/src/lib.rs#L1719-L1728

I don't believe this can be exported from `frame_support` however, as `frame_system` depends upon it.